gpt_index/data_structs/data_structs_v2.py

```diff
@@ -0,0 +1,379 @@
+"""Data structures v2.
+
+Nodes are decoupled from the indices.
+
+"""
+
+import random
+import sys
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from typing import Dict, Generic, List, Optional, Set, Tuple, TypeVar
+
+from dataclasses_json import DataClassJsonMixin
+
+from gpt_index.data_structs.data_structs import IndexGraph as V0IndexGraph
+from gpt_index.data_structs.data_structs import IndexStruct, Node
+from gpt_index.data_structs.struct_type import IndexStructType
+from gpt_index.schema import BaseDocument
+
+# OIS = TypeVar("IS", bound=IndexStruct)
+
+
+@dataclass
+class V2IndexStruct(IndexStruct, DataClassJsonMixin):
+    """A base data struct for a LlamaIndex."""
+
+    # # @abstractmethod
+    # @classmethod
+    # def from_v0_struct(cls, v0_struct: OIS) -> "V2IndexStruct":
+    #     """Convert from v0 struct."""
+    #     raise NotImplementedError()
+
+
+@dataclass
+class IndexGraph(V2IndexStruct):
+    """A graph representing the tree-structured index."""
+
+    # mapping from index in tree to Node doc id.
+    all_nodes: Dict[int, str] = field(default_factory=dict)
+    root_nodes: Dict[int, str] = field(default_factory=dict)
+
+    @property
+    def size(self) -> int:
+        """Get the size of the graph."""
+        return len(self.all_nodes)
+
+    def get_children(self, parent_node: Optional[Node]) -> Dict[int, str]:
+        """Get nodes given indices."""
+        if parent_node is None:
+            return self.root_nodes
+        else:
+            return {i: self.all_nodes[i] for i in parent_node.child_indices}
+
+    def insert_under_parent(self, node: Node, parent_node: Optional[Node]) -> None:
+        """Insert under parent node."""
+        if node.index in self.all_nodes:
+            raise ValueError(
+                "Cannot insert a new node with the same index as an existing node."
+            )
+        if parent_node is None:
+            self.root_nodes[node.index] = node.get_doc_id()
```

Comment:
Looks nodes are referenced via their str `doc_id` now. 

I find the naming slightly confusing, since a Node's `doc_id` is not the same as the original document `doc_id` (but a Node's `ref_doc_id` is the same as the original document `doc_id`).